dm/worker/log.go

```diff
@@ -0,0 +1,498 @@
+			break
+		}
+
+		if log.Id >= endPointer.Location {
+			// move to next location
+			endPointer.Location = log.Id + 1
+			logs = append(logs, log)
+		} else {
+			panic("out of sorted from level db")
+		}
+	}
+	iter.Release()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	err = iter.Error()
+	if err != nil {
+		return nil, errors.Annotatef(err, "fetch logs from meta")
+	}
+
+	logger.handledPointer = handledPointer
+	logger.endPointer = endPointer
+
+	log.Infof("handle pointer %+v, end pointer %+v", logger.handledPointer, logger.endPointer)
+
+	return logs, nil
+}
+
+// GetTaskLog returns task log by given log ID
+func (logger *Logger) GetTaskLog(h Getter, id int64) (*pb.TaskLog, error) {
+	if whetherNil(h) {
+		return nil, errors.Trace(ErrInValidHandler)
+	}
+
+	logBytes, err := h.Get(EncodeTaskLogKey(id), nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return nil, nil
+		}
+		return nil, errors.Annotatef(err, "get task meta from leveldb")
+	}
+
+	opLog := &pb.TaskLog{}
+	err = opLog.Unmarshal(logBytes)
+	if err != nil {
+		return nil, errors.Annotatef(err, "unmarshal task log binary %s", logBytes)
+	}
+
+	return opLog, nil
+}
+
+// ForwardTo forward handled pointer to specified ID location
+// not thread safe
+func (logger *Logger) ForwardTo(db Putter, ID int64) error {
+	if whetherNil(db) {
+		return errors.Trace(ErrInValidHandler)
+	}
+
+	handledPointer := Pointer{
+		Location: ID,
+	}
+
+	err := db.Put(HandledPointerKey, handledPointer.MarshalBinary(), nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	logger.handledPointer = handledPointer
+	return nil
+}
+
+// MarkAndForwardLog marks result sucess or not in log, and forwards handledPointer
+func (logger *Logger) MarkAndForwardLog(db Putter, opLog *pb.TaskLog) error {
+	if whetherNil(db) {
+		return errors.Trace(ErrInValidHandler)
+	}
+
+	logBytes, err := opLog.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Put(EncodeTaskLogKey(opLog.Id), logBytes, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(logger.ForwardTo(db, opLog.Id))
+}
+
+// Append appends a task log
+func (logger *Logger) Append(db Putter, opLog *pb.TaskLog) error {
+	if whetherNil(db) {
+		return errors.Trace(ErrInValidHandler)
+	}
+
+	var id int64
+	for {
+		id = atomic.LoadInt64(&logger.endPointer.Location)
+		if atomic.CompareAndSwapInt64(&logger.endPointer.Location, id, id+1) {
+			break
+		}
+	}
+
+	opLog.Id = id
+	opLog.Ts = time.Now().UnixNano()
+	logBytes, err := opLog.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Put(EncodeTaskLogKey(id), logBytes, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// GC deletes useless log
+func (logger *Logger) GC(ctx context.Context, db *leveldb.DB) {
+	ticker := time.NewTicker(time.Hour)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Infof("[worker] meta gc goroutine exist!")
```

Comment:
seems some log message not including `[worker]` mark in this file and some including other marks, is it expectant?

Reply:
we need to define a unify log format for this case